src/Avalonia.Base/Input/InputExtensions.cs

```diff
@@ -79,6 +78,53 @@ public static IEnumerable<IInputElement> GetInputElementsAt(this IInputElement e
         /// <inheritdoc cref="InputHitTest(IInputElement, Point, Func{Visual, bool}, bool)"/>
         public static IInputElement? InputHitTest(this IInputElement element, Point p, Func<Visual, bool> filter) => InputHitTest(element, p, filter, true);
 
+        internal static TEventArgs? RaiseEvent<TEventArgs, TContext>(
+            this IInputElement inputElement,
+            RoutedEvent<TEventArgs> routedEvent,
+            Func<RoutedEvent<TEventArgs>, TContext, TEventArgs> eventArgsFactory,
+            TContext context)
+            where TEventArgs : RoutedEventArgs
+        {
+            if (inputElement is InputElement typedInputElement)
+            {
+                return typedInputElement.RaiseEvent(routedEvent, eventArgsFactory, context);
+            }
+
+            var eventArgs = eventArgsFactory(routedEvent, context);
+            inputElement.RaiseEvent(eventArgs);
+            return eventArgs;
+        }
+
+        internal static TEventArgs? RaiseEvent<TEventArgs>(
+            this IInputElement inputElement,
+            RoutedEvent<TEventArgs> routedEvent,
+            Func<RoutedEvent<TEventArgs>, TEventArgs> eventArgsFactory)
+            where TEventArgs : RoutedEventArgs
+        {
+            if (inputElement is InputElement typedInputElement)
+            {
+                return typedInputElement.RaiseEvent(routedEvent, eventArgsFactory);
+            }
+
+            var eventArgs = eventArgsFactory(routedEvent);
+            inputElement.RaiseEvent(eventArgs);
+            return eventArgs;
+        }
+
+        internal static RoutedEventArgs? RaiseEvent(
+            this IInputElement inputElement,
+            RoutedEvent<RoutedEventArgs> routedEvent)
+        {
+            if (inputElement is InputElement typedInputElement)
+            {
+                return typedInputElement.RaiseEvent(routedEvent);
+            }
+
+            var eventArgs = new RoutedEventArgs(routedEvent);
```

Comment:
Won't it make sense to pass a factory here too?

Reply:
This would require adding all the new `RaiseEvent` overloads to the `IInputElement` interface, which I wasn't comfortable with, as I don't think it should be _that_ specific.

Since `IInputElement` isn't client implementable, we should never get to that fallback anyways.